core/src/main/java/com/linecorp/armeria/client/limit/ConcurrencyLimit.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.limit;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Integer.MAX_VALUE;
+
+import java.util.concurrent.CompletableFuture;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.util.SafeCloseable;
+
+/**
+ * A concurrency limiter based on an AsyncSemaphore.
```

Comment:
We could remove the implementation detail in the interface.

```suggestion
 * Limits the concurrency of client requests.
```